tests/common/auth_test.go

```diff
@@ -131,3 +140,287 @@ func TestAuthStatus(t *testing.T) {
+			t.Errorf("want error %s but got %v", rpctypes.ErrAuthFailed.Error(), err)
+		}
+	})
+}
+
+func TestAuthTxn(t *testing.T) {
+	tcs := []struct {
+		name string
+		cfg  config.ClusterConfig
+	}{
+		{
+			"NoJWT",
+			config.ClusterConfig{ClusterSize: 1},
+		},
+		{
+			"JWT",
+			config.ClusterConfig{ClusterSize: 1, AuthToken: defaultAuthToken},
+		},
+	}
+
+	reqs := []txnReq{
+		{
+			compare:   []string{`version("c2") = "1"`},
+			ifSuccess: []string{"get s2"},
+			ifFail:    []string{"get f2"},
+			results:   []string{"SUCCESS", "s2", "v"},
+		},
+		// a key of compare case isn't granted
+		{
+			compare:   []string{`version("c1") = "1"`},
+			ifSuccess: []string{"get s2"},
+			ifFail:    []string{"get f2"},
+			results:   []string{"etcdserver: permission denied"},
+		},
+		// a key of success case isn't granted
+		{
+			compare:   []string{`version("c2") = "1"`},
+			ifSuccess: []string{"get s1"},
+			ifFail:    []string{"get f2"},
+			results:   []string{"etcdserver: permission denied"},
+		},
+		// a key of failure case isn't granted
+		{
+			compare:   []string{`version("c2") = "1"`},
+			ifSuccess: []string{"get s2"},
+			ifFail:    []string{"get f1"},
+			results:   []string{"etcdserver: permission denied"},
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			testRunner.BeforeTest(t)
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+			clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.cfg))
+			defer clus.Close()
+			cc := testutils.MustClient(clus.Client())
+			testutils.ExecuteUntil(ctx, t, func() {
+				require.NoErrorf(t, setupAuth(cc, []authRole{testRole}, []authUser{rootUser, testUser}), "failed to enable auth")
+				rootAuthClient := testutils.MustClient(clus.Client(WithAuth(rootUserName, rootPassword)))
+				testUserAuthClient := testutils.MustClient(clus.Client(WithAuth(testUserName, testPassword)))
+
+				// keys with 1 suffix aren't granted to test-user
+				keys := []string{"c1", "s1", "f1"}
+				// keys with 2 suffix are granted to test-user, see Line 399
+				grantedKeys := []string{"c2", "s2", "f2"}
+				for _, key := range keys {
+					if err := cc.Put(ctx, key, "v", config.PutOptions{}); err != nil {
+						t.Fatal(err)
+					}
+				}
+				for _, key := range grantedKeys {
+					if err := cc.Put(ctx, key, "v", config.PutOptions{}); err != nil {
+						t.Fatal(err)
+					}
+				}
+
+				// grant keys to test-user
+				for _, key := range grantedKeys {
+					if _, err := rootAuthClient.RoleGrantPermission(ctx, testRoleName, key, "", clientv3.PermissionType(clientv3.PermReadWrite)); err != nil {
+						t.Fatal(err)
+					}
+				}
+				for _, req := range reqs {
+					resp, err := testUserAuthClient.Txn(ctx, req.compare, req.ifSuccess, req.ifFail, config.TxnOptions{
+						Interactive: true,
+					})
+					if strings.Contains(req.results[0], "denied") {
```

Comment:
Please add `expectError bool` to txnReq so we don't do use string matching here.